File: go/day3.go
package main

import (
	"main.go/go/helpers"
)

type item rune

func (i item) points() int {
	code := int(i)
	// lowercase
	if code >= 97 && code <= 122 {
		return code - 96
	}
	// uppercase
	if code >= 65 && code <= 90 {
		return code - 38
	}
	return 0
}

type rucksack map[item]bool

func (r rucksack) has(i item) bool {
	_, ok := r[i]
	return ok
}

func (r rucksack) add(i item) {
	r[i] = true
}

func (r rucksack) pop() (i item) {
	for it := range r {
		return it
	}
	return
}

func (r1 rucksack) intersection(r2 rucksack) rucksack {
	r3 := make(rucksack)
	for k := range r1 {
		if r2.has(k) {
			r3.add(k)
		}
	}
	return r3
}

type group []rucksack

func (g group) badge() item {
	rucksack := g[0]
	for i := 1; i < len(g); i++ {
		rucksack = rucksack.intersection(g[i])
	}
	return rucksack.pop()
}

func day3() (p1, p2 int) {
	file, scanner := helpers.GetFile(3)
	defer file.Close()

	g := group{}
	for scanner.Scan() {
		line := scanner.Text()
		r1, r2 := make(rucksack), make(rucksack)
		found := false
		m := len(line) / 2
		for i, c := range line {
			it := item(c)
			r2.add(it)
			if i < m {
				r1.add(it)
			} else if !found && r1.has(it) {
				found = true
				p1 += it.points()
			}
		}
		g = append(g, r2)
		if len(g) == 3 {
			p2 += g.badge().points()
			g = group{}
		}
	}

	return p1, p2
}

File: python/day3.py
class Item(str):
    @property
    def points(self):
        c = self[0]
        if c.lower() == c:
            return ord(c) - 96
        if c.upper() == c:
            return ord(c) - 38
        return 0


class Rucksack:
    def __init__(self, items):
        self.items = set([Item(i) for i in items])


class Group(list):
    @property
    def badge(self):
        items = set(self[0].items)
        for i in range(1, len(self)):
            items = items.intersection(self[i].items)
        return items.pop()


part_1_total = 0
part_2_total = 0
with open('data/day3.txt', 'r') as file:
    group = Group()
    for line in file:
        line = line.strip()
        middle = len(line) // 2

        # part1
        left, right = Rucksack(line[:middle]), Rucksack(line[middle:])
        part_1_total += left.items.intersection(right.items).pop().points

        # part2
        rucksack = Rucksack(left.items.union(right.items))
        group.append(rucksack)
        if len(group) == 3:
            part_2_total += group.badge.points
            group = Group()

print(part_1_total)
print(part_2_total)
